### PR TITLE
feat: add SystemCaller helper type

### LINKED ISSUE #10953: split apply preblock functions into two
### Describe the feature

we'd like to have this as two functions, one that commits, and one that doesn't

https://github.com/paradigmxyz/reth/blob/2a04b1c1002091172ff0db1e11d5c56cb80d53be/crates/evm/src/system_calls.rs#L172-L184

### Additional context

ref #10877

### LINKED ISSUE #10955: Add Helper type for system transactions
### Describe the feature

system transactions are repetitive, they use the same setup.

it'd be nice to have some parts reusable with a helper type and not just standalone functions.

Ideally a helper type that exposes functions that `apply` and `transact` the system calls.

this way we can chain the relevant calls in a nicer API, like:

```rust
SystemCalls::new(&mut evm, &evm_config, hook)
      .apply_blockhashes_contract_call(...)?
      .apply_beacon_root_contract_call(...)
      .finish()
```

the hook is for invoking the state output of the system calls, see `fn transact_*` calls after the system call, this way we can stream states (#11161):

https://github.com/paradigmxyz/reth/blob/da6b1e7c64417f346b3d9e0d36d2ccee2f13425e/crates/evm/src/system_calls/eip4788.rs#L71-L71

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
closes #10953 
closes #10955

Adds a helper type that can be used to chain system transactions.

This supports a draft for an optional hook to address: #10974

@Rjected

### CODE REVIEW

File: crates/evm/src/system_calls/eip7002.rs [hunk 1]
```diff
@@ -118,6 +118,11 @@ where
     // commit the state
     evm.context.evm.db.commit(state);
 
+    post_commit(result)
+}
+
+#[inline]
+pub(crate) fn post_commit(result: ExecutionResult) -> Result<Vec<Request>, BlockExecutionError> {
```
Comment (mattsse) [inline crates/evm/src/system_calls/eip7002.rs:125]: can we document here that this parses the requests from the execution output?

**Review by mattsse — COMMENTED** [review]
cool, only some suggestions

File: crates/evm/src/system_calls/eip7002.rs (hunk 1 above)
Comment (fgimenez) [reply crates/evm/src/system_calls/eip7002.rs:126]: sure, done here https://github.com/paradigmxyz/reth/pull/11068/commits/7cad4380d5af1112b954126d146b81161cb56f56

File: crates/evm/src/system_calls/mod.rs [hunk 2]
```diff
@@ -11,3 +20,314 @@ pub use eip7002::*;
...
+    fn on_state(&mut self, _state: &ResultAndState) {}
+}
+
+/// An ephemeral helper type for executing system calls.
+///
+/// This can be used to chain system transaction calls.
+#[allow(missing_debug_implementations)]
+pub struct SystemCaller<'a, EvmConfig, Chainspec, Hook = NoopHook> {
+    evm_config: &'a EvmConfig,
+    chain_spec: Arc<Chainspec>,
```
[the remark this answers is not in the document]
Comment (fgimenez) [reply crates/evm/src/system_calls/mod.rs]: sure done here https://github.com/paradigmxyz/reth/pull/11068/commits/011367c1043acc9229dbee56c7ca2c61c6b5a99b

File: crates/evm/src/system_calls/mod.rs [hunk 3]
```diff
@@ -11,3 +20,314 @@ pub use eip7002::*;
...
+impl OnStateHook for NoopHook {
+    fn on_state(&mut self, _state: &ResultAndState) {}
+}
+
+/// An ephemeral helper type for executing system calls.
+///
+/// This can be used to chain system transaction calls.
+#[allow(missing_debug_implementations)]
+pub struct SystemCaller<'a, EvmConfig, Chainspec, Hook = NoopHook> {
+    evm_config: &'a EvmConfig,
```
[the remark this answers is not in the document]
Comment (fgimenez) [reply crates/evm/src/system_calls/mod.rs:53]: without specifying the lifetime of `EvmConfig` we get errors like this:
```
error[E0521]: borrowed data escapes outside of method
   --> crates/ethereum/evm/src/execute.rs:142:33
    |
134 |         &self,
    |         -----
    |         |
    |         `self` is a reference that is only valid in the method body
    |         let's call the lifetime of this reference `'1`
...
142 |         let mut system_caller = SystemCaller::new(&self.evm_config, self.chain_spec.clone());
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                                 |
    |                                 `self` escapes the method body here
    |                                 argument requires that `'1` must outlive `'static`
```
agreed online to keep it as is.

**Review by fgimenez — APPROVED** [review]

**Pull request merged**